include/ginkgo/core/base/range_accessor_helper.hpp

```diff
@@ -0,0 +1,586 @@
+
+#ifndef GKO_CORE_BASE_RANGE_ACCESSOR_HELPER_HPP_
+#define GKO_CORE_BASE_RANGE_ACCESSOR_HELPER_HPP_
+
+#include <array>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/range.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+namespace accessor {
+/**
+ * This namespace contains helper functionality for the accessors.
+ *
+ * @note This namespace is not part of the public interface and can change
+ * without notice.
+ */
+namespace helper {
+
+
+namespace detail {
+
+
+/**
+ * This helper runs from first to last dimension in order to compute the index.
+ * The index is computed like this:
+ * indices: x1, x2, x3, ...
+ * compute(stride, x1, x2, x3) -> x1 * stride[0] + x2 * stride[1] + x3
+ */
+template <typename ValueType, size_type total_dim, size_type current_iter = 1>
+struct row_major_helper_s {
+    static_assert(total_dim >= 1, "Dimensionality must be >= 1");
+    static_assert(current_iter < total_dim, "Iteration must be < total_dim!");
+
+    static constexpr size_type dim_idx{current_iter - 1};
+
+    template <typename FirstType, typename... Indices>
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    compute(const dim<total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>
+                &stride,
+            FirstType &&first, Indices &&... idxs)
+    {
+        // The ASSERT size check must NOT be indexed with `dim_idx` directy,
+        // otherwise, it leads to a linker error. The reason is likely that
+        // `dim<N>::operator[](const size_type &)` uses a reference. Since
+        // `dim_idx` is constexpr (and not defined in a namespace scope), it
+        // can't be odr-used.
+        return GKO_ASSERT(first < size[static_cast<size_type>(dim_idx)]),
+               first * stride[dim_idx] +
+                   row_major_helper_s<ValueType, total_dim, current_iter + 1>::
+                       compute(size, stride, std::forward<Indices>(idxs)...);
+    }
+};
+
+template <typename ValueType, size_type total_dim>
+struct row_major_helper_s<ValueType, total_dim, total_dim> {
+    template <typename FirstType>
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    compute(const dim<total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>,
+            FirstType &&first)
+    {
+        return GKO_ASSERT(first < size[total_dim - 1]), first;
+    }
+};
+
+
+template <typename ValueType, size_type iter, size_type N>
+constexpr GKO_ATTRIBUTES GKO_INLINE std::enable_if_t<iter == N, ValueType>
+mult_array(const dim<N> &size)
+{
+    return 1;
+}
+
+template <typename ValueType, size_type iter, size_type N>
+constexpr GKO_ATTRIBUTES GKO_INLINE std::enable_if_t<(iter < N), ValueType>
+mult_array(const dim<N> &size)
+{
+    return size[iter] * mult_array<ValueType, iter + 1>(size);
+}
```

Comment:
Looking at the rest of the file,  lot of functions are `dim` specific (such as successively computing all strides, or in a masked fashion), and could be moved to the dim header as well. While they are needed by the accessor, I don't think they are accessor specific.

Reply:
I didn't see another use-case for this functionality when I implemented it (now, I renamed it to `mult_size_upwards`).
I agree I would not search for it in this place in case I ever needed it (and would not have written it myself). The main reason why I did not add it do `dim` is that I try to put as little code into a public header as possible.

Now, thinking about it, it is useful to compute the total number of elements for a given size that way. I will need to think about a better name for it, but I think it makes sense to move this into `dim.hpp`.

The masked stuff is something I would keep here since I don't see any other use case for it, and I don't know if there will ever be one outside of accessors.